File: src/Template/PackageInfoTemplate.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Net;
using System.Text;
using AutoRest.Core;
using AutoRest.Core.Utilities;
using AutoRest.Extensions;
using AutoRest.Core.Model;
using Newtonsoft.Json;
using AutoRest.Core.Utilities.Collections;
using Newtonsoft.Json.Linq;
using AutoRest.Java.Model;
using System.Text.RegularExpressions;

namespace AutoRest.Java
{
    public class PackageInfoTemplate : IJavaTemplate<PackageInfo, JavaFile>
    {
        private JavaSettings settings;
        private TemplateFactory factory;

        public PackageInfoTemplate(TemplateFactory factory)
        {
            this.factory = factory;
            this.settings = factory.Settings;
        }

        public void Write(PackageInfo packageInfo, JavaFile javaFile)
        {            
            if (!string.IsNullOrEmpty(settings.FileHeaderText))
            {
                javaFile.LineComment(settings.MaximumJavadocCommentWidth, (comment) =>
                {
                    comment.Line(settings.FileHeaderText);
                });
                javaFile.Line();
            }

            javaFile.JavadocComment(settings.MaximumJavadocCommentWidth, (comment) =>
            {
                foreach (var desc in packageInfo.Description.Split(new [] { '\r', '\n'}))
                {
                    comment.Description(desc);
                }
            });

            javaFile.Package(packageInfo.Package);
        }
    }
}

File: javagen/src/main/java/com/azure/autorest/template/PackageInfoTemplate.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.autorest.template;


import com.azure.autorest.extension.base.plugin.JavaSettings;
import com.azure.autorest.model.clientmodel.PackageInfo;
import com.azure.autorest.model.javamodel.JavaFile;

import java.util.regex.Pattern;

/**
 * Writes a PackageInfo to a JavaFile.
 */
public class PackageInfoTemplate implements IJavaTemplate<PackageInfo, JavaFile> {
    private static final PackageInfoTemplate INSTANCE = new PackageInfoTemplate();

    private static final Pattern NEW_LINE = Pattern.compile(Pattern.quote("\r\n"));

    private PackageInfoTemplate() {
    }

    public static PackageInfoTemplate getInstance() {
        return INSTANCE;
    }

    public final void write(PackageInfo packageInfo, JavaFile javaFile) {
        JavaSettings settings = JavaSettings.getInstance();
        if (settings.getFileHeaderText() != null && !settings.getFileHeaderText().isEmpty()) {
            javaFile.lineComment(settings.getMaximumJavadocCommentWidth(), (comment) ->
                comment.line(settings.getFileHeaderText()));
            javaFile.line();
        }

        javaFile.javadocComment(settings.getMaximumJavadocCommentWidth(), (comment) -> {
            for (String desc : NEW_LINE.split(packageInfo.getDescription(), -1)) {
                comment.description(desc);
            }
        });

        javaFile.declarePackage(packageInfo.getPackage());
    }
}
